gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/planning/planner/greedy/GreedyPlanner.java

```diff
@@ -0,0 +1,394 @@
+      Set<String> projectionKeys = allPredicates.getPropertyKeys(edgeVariable);
+
+      FilterAndProjectEdgesNode node = new FilterAndProjectEdgesNode(graph.getEdges(),
+        sourceVariable, edgeVariable, targetVariable, edgePredicates, projectionKeys);
+
+      PlanTableEntry.Type type = edge.hasVariableLength() ? PATH : EDGE;
+
+      planTable.add(new PlanTableEntry(type, Sets.newHashSet(edgeVariable), allPredicates,
+        new QueryPlanEstimator(new QueryPlan(node), queryHandler, graphStatistics)));
+    }
+    return planTable;
+  }
+
+  //------------------------------------------------------------------------------------------------
+  // Join and Expand
+  //------------------------------------------------------------------------------------------------
+
+  /**
+   * Evaluates which entries in the specified plan table can be joined. The joined entries
+   * are added to a new table which is returned.
+   *
+   * @param currentTable query plan table
+   * @return new table containing solely joined plans from the input table
+   */
+  private PlanTable evaluateJoins(PlanTable currentTable) {
+    PlanTable newTable = new PlanTable();
+
+    for (int i = 0; i < currentTable.size(); i++) {
+      PlanTableEntry leftEntry = currentTable.get(i);
+      if (mayExtend(leftEntry)) {
+        for (int j = 0; j < currentTable.size(); j++) {
+          PlanTableEntry rightEntry = currentTable.get(j);
+          if (i != j) {
+            List<String> joinVariables = getOverlap(leftEntry, rightEntry);
+            if (joinVariables.size() > 0) {
+              if (rightEntry.getType() == PATH && joinVariables.size() == 2) {
+                // evaluate join with variable length path on source and target vertex
+                newTable.add(joinEntries(leftEntry, rightEntry, joinVariables.subList(0, 1)));
+                newTable.add(joinEntries(leftEntry, rightEntry, joinVariables.subList(1, 2)));
+              } else {
+                // regular join or join with variable length path on source or target vertex
+                newTable.add(joinEntries(leftEntry, rightEntry, joinVariables));
+              }
+            }
+          }
+        }
+      }
+    }
+    return newTable;
+  }
+
+  /**
+   * Checks if the given entry may be extended. This is only the case for entries that represents
+   * either a vertex or a partial match graph.
+   *
+   * @param entry plan table entry
+   * @return true, iff the specified entry may be extended
+   */
+  private boolean mayExtend(PlanTableEntry entry) {
+    return entry.getType() == VERTEX || entry.getType() == GRAPH;
+  }
+
+  /**
+   * Computes the overlapping query variables of the specified entries.
+   *
+   * @param firstEntry first entry
+   * @param secondEntry second entry
+   * @return variables that are available in both input entries
+   */
+  private List<String> getOverlap(PlanTableEntry firstEntry, PlanTableEntry secondEntry) {
+    return firstEntry.getAllVariables().stream()
```

Comment:
Could we convert getAllVariables to return a set? --> would make it possible to use `Set.retainAll(Collection other)`